CRM/Contribute/Tokens.php

```diff
@@ -96,35 +115,122 @@ public function alterActionScheduleQuery(\Civi\ActionSchedule\Event\MailingQuery
       return;
     }
 
+    $passThroughTokens = [
+      'contribution_page_id',
+      'receive_date',
+      'total_amount',
+      'fee_amount',
+      'net_amount',
+      'trxn_id',
+      'invoice_id',
+      'currency',
+      'contribution_cancel_date',
+      'receipt_date',
+      'thankyou_date',
+      'tax_amount',
+    ];
     $fields = CRM_Contribute_DAO_Contribution::fields();
-    foreach ($this->getPassthruTokens() as $token) {
+    foreach ($passThroughTokens as $token) {
       $e->query->select("e." . $fields[$token]['name'] . " AS contrib_{$token}");
     }
-    foreach ($this->getAliasTokens() as $alias => $orig) {
+    $aliasTokens = [
+      'id' => 'contribution_id',
+      'payment_instrument' => 'payment_instrument_id',
+      'source' => 'contribution_source',
+      'status' => 'contribution_status_id',
+      'type' => 'financial_type_id',
+      'cancel_date' => 'contribution_cancel_date',
+    ];
+    foreach ($aliasTokens as $alias => $orig) {
       $e->query->select("e." . $fields[$orig]['name'] . " AS contrib_{$alias}");
     }
   }
 
   /**
    * @inheritDoc
    */
-  public function evaluateToken(\Civi\Token\TokenRow $row, $entity, $field, $prefetch = NULL) {
-    $actionSearchResult = $row->context['actionSearchResult'];
-    $fieldValue = $actionSearchResult->{"contrib_$field"} ?? NULL;
+  public function prefetch(TokenValueEvent $e) {
+    // Find all the entity IDs
+    $entityIds
+      = $e->getTokenProcessor()->getContextValues('actionSearchResult', 'entityID')
+      + $e->getTokenProcessor()->getContextValues($this->getEntityContextSchema());
+
+    if (!$entityIds) {
+      return NULL;
+    }
+
+    $returnFields = self::getReturnFields($this->activeTokens);
+    if (!in_array('currency', $returnFields)) {
+      // We need currency if we have an amount
+      $returnFields[] = 'currency';
+    }
+    // Get data on all entities for basic and customfield tokens
+    $entities = civicrm_api3('Contribution', 'get', [
+      'id' => ['IN' => $entityIds],
+      'options' => ['limit' => 0],
+      'return' => $returnFields,
+    ]);
+    $prefetch['contribution'] = $entities['values'];
```

Comment:
Aren't we preferring APIv4 wherever possible?

Reply:
Yep, replaced with API4 equivalent